File: cs/src/AutoDiff/Variable.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AutoDiff
{
    /// <summary>
    /// Represents a variable term. Variable terms are substituted for real values during evaluation and
    /// differentiation. 
    /// </summary>
    [Serializable]
    public class Variable : Term
    {
        /// <summary>
        /// Accepts a term visitor
        /// </summary>
        /// <param name="visitor">The term visitor to accept.</param>
        public override void Accept(ITermVisitor visitor)
        {
            visitor.Visit(this);
        }

        /// <summary>
        /// Accepts a term visitor with a generic result
        /// </summary>
        /// <typeparam name="TResult">The type of the result from the visitor's function</typeparam>
        /// <param name="visitor">The visitor to accept</param>
        /// <returns>
        /// The result from the visitor's visit function.
        /// </returns>
        public override TResult Accept<TResult>(ITermVisitor<TResult> visitor)
        {
            return visitor.Visit(this);
        }
		/// Additions by Carpe Noctem:
		
		public double GlobalMin = Double.NegativeInfinity;
		public double GlobalMax = Double.PositiveInfinity;
		static int id=0;
		int ownID;
		
		public Variable() {
			ownID = id++;
		}
		
		public override string ToString()
		{
			int hash = ownID;
			//return (hash<0 ? string.Format ("Var_{0}[{1}..{2}]",-hash,Min,Max) : string.Format ("Var{0}[{1}..{2}]",hash,Min,Max));
			return (hash<0 ? string.Format ("Var_{0}",-hash) : string.Format ("Var{0}",hash));
		}
		
		public override Term AggregateConstants()
		{
			return this;
		}
		public override Term Derivative(Variable v)
		{
			if(this==v) return 1;
			else return 0;
		}
    }
}

File: java/src/main/java/AutoDiff/Variable.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:34
//

package AutoDiff;

/**
* Represents a variable term. Variable terms are substituted for real values during evaluation and
* differentiation.
*/
public class Variable  extends Term 
{
//    /**
//    * Accepts a term visitor
//    *
//    *  @param visitor The term visitor to accept.
//    */
//    public void accept(ITermVisitor visitor) throws Exception {
//        visitor.visit(this);
//    }

    /**
    * Accepts a term visitor with a generic result
    * The type of the result from the visitor's function
    *  @param visitor The visitor to accept
    *  @return 
    * The result from the visitor's visit function.
    */
    public <TResult>TResult accept(ITermVisitor<TResult> visitor) throws Exception {
        return visitor.visit(this);
    }

    /**
    * Additions by Carpe Noctem:
    */
    public double GlobalMin = Double.NEGATIVE_INFINITY;
    public double GlobalMax = Double.POSITIVE_INFINITY;

    static int id = 0;
    String name="Var";
    int ownID;
    public Variable() throws Exception {
        ownID = id++;
    }

    public String toString() {
        try
        {
            int hash = ownID;
            return (hash < 0 ? String.format("%s_%d",name, -hash) : String.format("%s%d", name, hash));
        }
        catch (RuntimeException exception)
        {
            throw exception;
        }
        catch (Exception exception)
        {
            throw new RuntimeException(exception);
        }
    
    }

    //return (hash<0 ? string.Format ("Var_{0}[{1}..{2}]",-hash,Min,Max) : string.Format ("Var{0}[{1}..{2}]",hash,Min,Max));
    public Term aggregateConstants() throws Exception {
        return this;
    }

    public Term derivative(Variable v) throws Exception {
        if (this == v)
            return TermBuilder.Constant(1);
        else
            return TermBuilder.Constant(0);
    }

    public String getName()
    {
        return name;
    }

    public void setName(final String name)
    {
        this.name = name;
    }

    public static void resetIdCounter() {
        id = 0;
    }
}


